velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -34,33 +103,138 @@ HiveDataSink::HiveDataSink(
     : inputType_(std::move(inputType)),
       insertTableHandle_(std::move(insertTableHandle)),
       connectorQueryCtx_(connectorQueryCtx),
-      writeProtocol_(std::move(writeProtocol)) {
+      writeProtocol_(std::move(writeProtocol)),
+      partitionChannels_(getPartitionChannels(insertTableHandle_)),
+      partitionsType_(getChannelTypes(inputType_, partitionChannels_)),
+      partitionIdGenerator_(partitionsType_, partitionChannels_) {
   VELOX_CHECK_NOT_NULL(
       writeProtocol_, "Write protocol could not be nullptr for HiveDataSink.");
+
+  partitionsVector_ = std::static_pointer_cast<RowVector>(
+      BaseVector::create(partitionsType_, 0, connectorQueryCtx_->memoryPool()));
 }
 
 std::shared_ptr<ConnectorCommitInfo> HiveDataSink::getConnectorCommitInfo()
     const {
-  return std::make_shared<HiveConnectorCommitInfo>(writerParameters_);
+  std::vector<std::shared_ptr<const HiveWriterParameters>> toCommit;
+  toCommit.reserve(writerParameters_.size());
+
+  for (auto i = 0; i < writerParameters_.size(); i++) {
+    if (writerParameters_[i] != nullptr) {
+      toCommit.push_back(writerParameters_[i]);
+    }
+  }
+  return std::make_shared<HiveConnectorCommitInfo>(toCommit);
 }
 
 void HiveDataSink::appendData(VectorPtr input) {
-  // For the time being the hive data sink supports one file
-  // To extend it we can create a new writer for every
-  // partition
-  if (writers_.empty()) {
-    writers_.emplace_back(createWriter());
+  input_ = std::dynamic_pointer_cast<RowVector>(input);
+  input_->loadedVector();
```

Comment:
input is never a lazy vector. Hence, calling loadedVector is redundant. However, child vectors might be lazy. Hence, you need to call loadedVector on them.

Here is an example from HashProbe.cpp:

```
  // Ensure vector are lazy loaded before spilling.
  for (int32_t i = 0; i < input_->childrenSize(); ++i) {
    input_->childAt(i)->loadedVector();
  }
```

Reply:
Got you, good to know. I change to call loadedVector() on individual child vectors of input_, and it works!